File: web/src/tracking/handGestureDetector.ts
import type { DetectedHandGesture, DetectedHandSide } from './gestureRecognizerClient'

export type HandSide = DetectedHandSide

export type HandPose = 'open' | 'closed' | null

export interface HandGestureState {
  leftPose: HandPose
  rightPose: HandPose
}

const OPEN_GESTURE = 'Open_Palm'
const CLOSED_GESTURE = 'Closed_Fist'

export interface HandGestureDetectorConfig {
  minScore: number
  stableFrames: number
  cooldownMs: number
}

interface SideTracker {
  stablePose: HandPose
  pendingPose: HandPose
  pendingCount: number
}

/** Detects open↔closed hand transitions for left/right phrase selection. */
export class HandGestureDetector {
  private left: SideTracker = this.freshSide()
  private right: SideTracker = this.freshSide()
  private lastActivationTime = 0

  reset(): void {
    this.left = this.freshSide()
    this.right = this.freshSide()
  }

  process(
    hands: DetectedHandGesture[],
    now: number,
    config: HandGestureDetectorConfig,
  ): { activation: HandSide | null; state: HandGestureState } {
    const detected: Partial<Record<HandSide, HandPose>> = {}

    for (const hand of hands) {
      const pose = this.gestureToPose(hand.gestureName, hand.score, config.minScore)
      if (pose) detected[hand.side] = pose
    }

    let activation: HandSide | null = null

    if (now - this.lastActivationTime >= config.cooldownMs) {
      activation = this.advanceSide('left', this.left, detected.left ?? null, config.stableFrames)
      if (!activation) {
        activation = this.advanceSide('right', this.right, detected.right ?? null, config.stableFrames)
      }
      if (activation) this.lastActivationTime = now
    } else {
      this.advanceSide('left', this.left, detected.left ?? null, config.stableFrames)
      this.advanceSide('right', this.right, detected.right ?? null, config.stableFrames)
    }

    return {
      activation,
      state: {
        leftPose: this.left.stablePose,
        rightPose: this.right.stablePose,
      },
    }
  }

  private freshSide(): SideTracker {
    return { stablePose: null, pendingPose: null, pendingCount: 0 }
  }

  private gestureToPose(name: string, score: number, minScore: number): HandPose {
    if (score < minScore) return null
    if (name === OPEN_GESTURE) return 'open'
    if (name === CLOSED_GESTURE) return 'closed'
    return null
  }

  private advanceSide(
    side: HandSide,
    tracker: SideTracker,
    pose: HandPose,
    stableFrames: number,
  ): HandSide | null {
    if (pose == null) {
      tracker.pendingPose = null
      tracker.pendingCount = 0
      return null
    }

    if (pose !== tracker.pendingPose) {
      tracker.pendingPose = pose
      tracker.pendingCount = 1
      return null
    }

    tracker.pendingCount++
    if (tracker.pendingCount < stableFrames) return null

    const previous = tracker.stablePose
    if (previous != null && previous !== pose) {
      tracker.stablePose = pose
      tracker.pendingPose = null
      tracker.pendingCount = 0
      return side
    }

    tracker.stablePose = pose
    return null
  }
}

File: iosApp/iosApp/Tracking/HandGestureDetector.swift
import Foundation

typealias HandSide = ArmSide

enum HandPose: Equatable {
    case open
    case closed
}

struct HandGestureState: Equatable {
    var leftPose: HandPose?
    var rightPose: HandPose?
}

struct DetectedHandGesture {
    let side: HandSide
    let gestureName: String
    let score: Float
}

struct HandGestureDetectorConfig {
    var minScore: Float
    var stableFrames: Int
    var cooldownMs: Double
}

/// Detects open↔closed hand transitions for left/right phrase selection.
final class HandGestureDetector {
    private struct SideTracker {
        var stablePose: HandPose?
        var pendingPose: HandPose?
        var pendingCount: Int
    }

    private static let openGesture = "Open_Palm"
    private static let closedGesture = "Closed_Fist"

    private var left = SideTracker(stablePose: nil, pendingPose: nil, pendingCount: 0)
    private var right = SideTracker(stablePose: nil, pendingPose: nil, pendingCount: 0)
    private var lastActivationTime: Double = 0

    func reset() {
        left = freshSide()
        right = freshSide()
    }

    func process(
        hands: [DetectedHandGesture],
        now: Double,
        config: HandGestureDetectorConfig
    ) -> (activation: HandSide?, state: HandGestureState) {
        var detected: [HandSide: HandPose] = [:]
        for hand in hands {
            if let pose = gestureToPose(name: hand.gestureName, score: hand.score, minScore: config.minScore) {
                detected[hand.side] = pose
            }
        }

        var activation: HandSide?
        if now - lastActivationTime >= config.cooldownMs {
            activation = advanceSide(.left, tracker: &left, pose: detected[.left], stableFrames: config.stableFrames)
            if activation == nil {
                activation = advanceSide(.right, tracker: &right, pose: detected[.right], stableFrames: config.stableFrames)
            }
            if activation != nil {
                lastActivationTime = now
            }
        } else {
            _ = advanceSide(.left, tracker: &left, pose: detected[.left], stableFrames: config.stableFrames)
            _ = advanceSide(.right, tracker: &right, pose: detected[.right], stableFrames: config.stableFrames)
        }

        return (
            activation,
            HandGestureState(leftPose: left.stablePose, rightPose: right.stablePose)
        )
    }

    private func freshSide() -> SideTracker {
        SideTracker(stablePose: nil, pendingPose: nil, pendingCount: 0)
    }

    private func gestureToPose(name: String, score: Float, minScore: Float) -> HandPose? {
        guard score >= minScore else { return nil }
        if name == Self.openGesture { return .open }
        if name == Self.closedGesture { return .closed }
        return nil
    }

    private func advanceSide(
        _ side: HandSide,
        tracker: inout SideTracker,
        pose: HandPose?,
        stableFrames: Int
    ) -> HandSide? {
        guard let pose else {
            tracker.pendingPose = nil
            tracker.pendingCount = 0
            return nil
        }

        if pose != tracker.pendingPose {
            tracker.pendingPose = pose
            tracker.pendingCount = 1
            return nil
        }

        tracker.pendingCount += 1
        if tracker.pendingCount < stableFrames { return nil }

        let previous = tracker.stablePose
        if previous != nil && previous != pose {
            tracker.stablePose = pose
            tracker.pendingPose = nil
            tracker.pendingCount = 0
            return side
        }

        tracker.stablePose = pose
        return nil
    }
}
